<!-- EBI technical test 2018. By Robin Dunn. -->
<html>

    <head>
        <style>
            body { 
                font-family: Arial, Helvetica, sans-serif; 
                line-height: 48px; font-size: 20px
            }
            p { margin-bottom: 0; }
            table { 
                position:relative; display:table; width:100%; 
                border-collapse: collapse; table-layout: fixed; 
            }
            #page { overflow:auto; margin: 0 auto; width: 600px; }
            #appTitle { font-size:24px; font-weight:normal; text-align: center }
            .form fieldset {
                border:2px solid #aaa;
            }
            .form input[type=text]{
                outline: 0;
                font-size: 20px; line-height:32px; 
                padding: 0 6px; margin: 0 0 10px 0;
                border: none; border-bottom: 1px solid #ccc;
            }
            #inputGeneSymbol { width: 110px } 
            #inputSeqPosition { width: 90px } 
            #inputAminoAcid { width:40px; }
            #inputHgvs { width: 280px; }
            #btnSearchSimple {  }
            #btnSearchHgvs {  }
            .btnSearch { 
                outline: 0;
                font-size: 20px; 
                line-height: 32px;
                padding: 0 20px;
                border:1px solid #ccc; 
                background-color: #fff;
            }
            #results { margin:20px 0 0 0; }
            th,td { text-align:center; padding: 0 10px; line-height: 32px; }
            th { font-weight: normal; border-bottom: 5px solid #ccc }
        </style>
    </head>

    <body>
        <div id="page">
            <h1 id="appTitle">EBI Technical Test - REST Client</h1>
            <div class='form'>
                <div>
                    <fieldset>
                        <legend>Enter gene symbol, sequence position and amino acid</legend>
                        <input id='inputGeneSymbol' title='Gene symbol' type='text' 
                            maxlength='50' onkeyup='onFieldChange(event);'/>
                        <input id='inputSeqPosition' title='Sequence position' type='text' 
                            maxlength="6" onkeyup='onFieldChange(event);'/>
                        <input id='inputAminoAcid' title='Amino acid letter' type='text' 
                            maxlength="6" onkeyup='onFieldChange(event);'/>
                        <button id='btnSearchSimple' class='btnSearch' 
                            onclick="onSearchClick(event, 'simple');">Search</button>
                    </fieldset>
                    <fieldset>
                        <legend>Enter HGVS notation</legend>
                        <!-- <REFERENCE_SEQUENCE_ID>:<SEQUENCE_TYPE>.<POSITION><CHANGE> -->
                        <input id='inputHgvs' type='text' maxlength="70" 
                            placeholder='Symbol:Type.PositionChange' onkeyup='onFieldChange(event);'/>
                        <button id='btnSearchHgvs' class='btnSearch' 
                            onclick="onSearchClick(event, 'hgvs');">Search</button>
                    </fieldset>
                </div>
                <div id='searching' style='display:none;'>Searching</div>
                <div id='results' style='display:none;'>
                    <p id='totalResults'></p>
                    <!-- Table of matching transcripts -->
                    <table id='resultsTable'>
                        <thead>
                            <th>ID</th><th>Name</th><th>Link</th>
                        </thead>
                        <tbody></tbody>
                    </table>
                </div>
            </div>
        </div>

        <script src='Extensions.js'></script>
        <script src='Http.js'></script>
        <script src='State.js'></script>
        <script src='EnsemblApi.js'></script>
        <script src='GeneTools.js'></script>
        <script>
            var _ = Extensions;
            var state = State;
            /* ///////////////////////////////////////////////////////////////// */
            // Main render function updates DOM according to state.
            /* ///////////////////////////////////////////////////////////////// */
            var render = function(state){
                //console.log('RENDER', state);
                _.$('inputGeneSymbol').value = state.geneSymbol;
                _.$('inputSeqPosition').value = state.seqPosition;
                _.$('inputAminoAcid').value = state.aminoAcid;
                _.$('inputHgvs').value = state.hgvsQuery;
                _.$('searching').style.display = state.searching ? 'inline-block' : 'none';
                var transcripts = [];
                for(var i = 0; i < state.matchedTranscripts.length; i++){
                    var ts = state.matchedTranscripts[i];
                    if(typeof ts === 'object') transcripts.push(ts);
                }
                var results = _.$('results');
                var resultsTable = _.$('resultsTable');
                if(state.searchDone && _.isArray(transcripts) && transcripts.length >= 1){
                    results.style.display = 'block'; 
                    _.$('totalResults').innerHTML = 'Found '+transcripts.length + ' matches.';
                    // Build the table rows
                    var newTbody = document.createElement('tbody');
                    var oldTbody = resultsTable.querySelectorAll('tbody')[0];
                    resultsTable.replaceChild(newTbody, oldTbody);
                    // Insert table row for each match found
                    for(var i = 0; i < transcripts.length; i++){
                        var ts = transcripts[i];
                        var row = newTbody.insertRow(i);
                        var ensemblUrl = 'http://www.ensembl.org/id/'+ts.id;
                        var link = '<a target="blank" href="'+ensemblUrl+'">Ensemble</a>';
                        var c1 = row.insertCell(0); c1.innerHTML = ''+ts.id;
                        var c2 = row.insertCell(1); c2.innerHTML = ''+ts.display_name;
                        var c2 = row.insertCell(2); c2.innerHTML = link;
                    }
                } else {
                    results.style.display = 'none'; 
                }
            }
            state.onChangeCallback = render;
            //state.onChangeCallback = function(state){ console.log(state); };
            // Set default state
            state.set({ 
                geneSymbol: 'BRCA2', seqPosition: 1, aminoAcid: 'S',
                hgvsQuery:'BRAF.p:A600K', 
                requests:[], seqQueries: [], searchDone: false, searching: false, 
                searchParams: { 
                    method:'', symbol:'', type:'', 
                    position:0, aminoAcid:'', aminoAcidTo:'' 
                },
                geneInfo: '', matchedTranscripts: [], sequences: [] 
            });
            // Handle field change events and change state accordingly
            var onFieldChange = function(e){
                var newValue = e.target.value;
                switch(e.target.id){
                    case 'inputGeneSymbol': state.set({ geneSymbol: newValue }); break;
                    case 'inputSeqPosition': state.set({ seqPosition: newValue }); break;
                    case 'inputAminoAcid': state.set({ aminoAcid: newValue }); break;
                    case 'inputHgvs': state.set({ hgvsQuery: newValue }); break;
                    default: break;
                }
            }
            var doSearch = function(params){
                var inputSymbolOk = typeof params.symbol === 'string' && params.symbol.length >= 1;
                if(!inputSymbolOk) return;
                state.set({ 
                    searchDone: false, searching: true, geneInfo: null, 
                    requests: [], sequences: [], matchedTranscripts: [] 
                    }, 
                    function(){ 
                        EnsemblApi.getGeneBySymbol(params.symbol, 
                            function(respGene){ 
                                // Get reference sequence for the gene
                                EnsemblApi.getSequence(respGene.id, false, null, 
                                    function(respSeq){
                                        respGene.seq = respSeq.seq;
                                        state.set({ geneInfo: respGene }); 
                                        // Get sequences for the gene's transcripts
                                        if(!_.isArray(respGene.Transcript)) return;
                                        var transcripts = respGene.Transcript;
                                        for(var i = 0; i < transcripts.length; i++){
                                            if(transcripts[i].id == respGene.id) 
                                                continue; // Ignore reference sequence
                                            var req = EnsemblApi.getSequence(transcripts[i].id, true, 'protein',
                                                function(resp, status, xhr){
                                                    var sequences = state.get('sequences');
                                                    for(var i=0; i < resp.length; i++){
                                                        sequences.push( resp[i] );
                                                    }
                                                    state.set({ sequences: sequences }); 
                                                }
                                            );
                                            var requests = state.get('requests');
                                            requests.push(req);
                                            state.set({requests:requests});
                                        }
                                    }
                                );
                            }
                        );
                    }
                );
            }
            // Function to send ajax search request
            var onSearchClick = function(e, searchType){
                if(searchType != 'simple' && searchType != 'hgvs') searchType = 'simple';
                var params = state.get('searchParams');
                params.method = searchType;
                switch(searchType.toLowerCase()){
                    case 'simple': 
                        params.symbol = state.get('geneSymbol');
                        params.type = 'protein';
                        params.position = state.get('seqPosition');
                        params.aminoAcid = state.get('aminoAcid');
                        break;
                    case 'hgvs': 
                        var query = state.get('hgvsQuery');
                        var parts = GeneTools.parseHgvsQuery(query);
                        params.symbol = parts.symbol;
                        params.type = parts.type;
                        var strChange = parts.positionChange;
                        if(strChange.length >= 1) 
                            params.aminoAcid = strChange.charAt(0);
                        if(strChange.length >= 2) 
                            params.position = strChange.replace(/[^0-9]/g,'');
                        if(strChange.length >= 3) 
                            params.aminoAcidTo = strChange.charAt(strChange.length - 1);
                        break;
                    default: break;
                }
                // Run the search
                state.set({ searchParams: params }, function(){ doSearch(params); });
            };
            
            function getMatchedTranscripts(gene, searchParams, allTranscripts, sequences){
                var matchedTranscripts = [];
                if(searchParams.method === 'simple'){
                    for(var i = 0; i < sequences.length; i++){
                        var sequence = sequences[i];
                        var intPos = parseInt(searchParams.position);
                        intPos = intPos >= 0 ? intPos : 0;
                        if(GeneTools.hasCharAt(sequence.seq, intPos, searchParams.aminoAcid)){
                            // Find transcripts for this sequence
                            if(_.find(matchedTranscripts, 'id', sequence.query) == null){
                                var matched = _.find(allTranscripts, 'id', sequence.query);
                                matchedTranscripts.push(matched);
                            }
                        }
                    }
                }
                if(searchParams.method == 'hgvs'){
                    // Check if change from reference sequence
                    searchParams.position = parseInt(searchParams.position);
                    if(_.isLetter(searchParams.aminoAcid) && _.isLetter(searchParams.aminoAcidTo)){
                        for(var i = 0; i < sequences.length; i++){
                            var s = sequences[i];
                            if(GeneTools.hasCharAt(gene.seq, searchParams.position, searchParams.aminoAcid)
                                && GeneTools.hasCharAt(s.seq, searchParams.position, searchParams.aminoAcidTo)){
                                // Find transcripts for this sequence
                                if(_.find(matchedTranscripts, 'id', s.query) == null){
                                    var matched = _.find(allTranscripts, 'id', s.query);
                                    matchedTranscripts.push(matched);
                                }
                            }
                        }
                    }
                }
                console.log('Matched transcripts: ', matchedTranscripts);
                return matchedTranscripts;                
            }

            window.onload = function(e){
                state.subscribe('sequences', function(sequences){
                    var requests = state.get('requests');
                    if(requests.length >= 1 && _.filter(requests, 'readyState', 4).length == requests.length){
                        console.log('All requests finished.', requests.length);
                        var gene = state.get('geneInfo');
                        console.log('Gene: ', gene);
                        if(typeof gene !== 'object' || gene == null) return;
                        var allTranscripts = gene.Transcript;
                        var searchParams = state.get('searchParams');
                        var matched = getMatchedTranscripts(gene, searchParams, allTranscripts, sequences);
                        state.set({ searchDone: true, searching: false, matchedTranscripts: matched });
                    }
                    return;
                });
            }
        </script>
    </body>

</html>
